lib/mongo_mapper/eager_includer.rb

```diff
@@ -1,194 +1,165 @@
-
-  private
-
-    def cache
-      @cache ||= {}
+  def self.eager_include(record_or_records, *association_names, &query_alteration_block)
+    association_names.each do |association_name|
+      new(record_or_records, association_name, &query_alteration_block).eager_include
     end
   end
 
-  def initialize(record_or_records, association_name)
-    association_name = association_name.to_sym
-
-    @records = Array(record_or_records)
-
-    if @records.length == 0
-      return
+  def initialize(record_or_records, association_name, &query_alteration_block)
+    if record_or_records.is_a? Plucky::Query
+      raise "You must call `to_a` on `Plucky::Query` objects before passing to eager_include"
     end
+    @records = Array(record_or_records)
 
+    return if @records.length == 0
     @association_name = association_name.to_sym
-    @association = @records.first.associations[association_name]
+    @query_alteration_block = query_alteration_block
+    @association = @records.first.associations[@association_name]
     if !@association
       raise "Could not find association `#{association_name}` on instance of #{@records.first.class}"
     end
-
-    @proxy_class = @association.proxy_class
   end
 
-  def enabled?
-    self.class.enabled?
-  end
-
-  def eager_include(&block)
-    return if !enabled?
-
-    if @records.length == 0
-      return
+  def eager_include
+    @records.reject! do |record|
+      record.instance_variable_defined?(instance_variable_name)
     end
 
-    if @proxy_class == MongoMapper::Plugins::Associations::ManyDocumentsProxy
-      eager_include_has_many(&block)
-    elsif @proxy_class == MongoMapper::Plugins::Associations::BelongsToProxy
-      eager_include_belongs_to(&block)
-    elsif @proxy_class == MongoMapper::Plugins::Associations::OneProxy
-      eager_include_has_one(&block)
-    elsif @proxy_class == MongoMapper::Plugins::Associations::InArrayProxy
-      eager_include_has_many_in(&block)
+    return if @records.length == 0
+
+    @association_type = case @association.proxy_class.to_s
+    when 'MongoMapper::Plugins::Associations::ManyDocumentsProxy'
+      :has_many
+    when 'MongoMapper::Plugins::Associations::BelongsToProxy'
+      :belongs_to
+    when 'MongoMapper::Plugins::Associations::OneProxy'
+      :has_one
+    when 'MongoMapper::Plugins::Associations::InArrayProxy'
+      :has_many_in
     else
-      raise NotImplementedError, "#{@proxy_class} not supported yet!"
+      raise NotImplementedError, "#{@association.proxy_class} not supported yet!"
     end
+
+    send("eager_include_#{@association_type}")
   end
 
 private
 
-  def setup_association(record, association_name, value)
-    association_name = association_name.to_sym
+  attr_reader :association_name
 
-    self.class.write_to_cache(record.object_id, association_name, value)
+  def instance_variable_name
+    "@eager_loaded_#{association_name}"
+  end
 
-    code = <<-CODE
+  def setup_association(record, value)
+    code = <<-RUBY
```

Comment:
`__FILE__, __LINE__` happens later when you eval the code. this just defines a string